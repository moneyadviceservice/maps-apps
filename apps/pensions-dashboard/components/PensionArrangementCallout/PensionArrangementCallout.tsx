import { Button } from '@maps-react/common/components/Button';
import { ExpandableSection } from '@maps-react/common/components/ExpandableSection';
import { Heading } from '@maps-react/common/components/Heading';
import { Icon, IconType } from '@maps-react/common/components/Icon';
import { InformationCallout } from '@maps-react/common/components/InformationCallout';
import { Paragraph } from '@maps-react/common/components/Paragraph';
import useTranslation from '@maps-react/hooks/useTranslation';

import { PensionType } from '../../lib/constants';
import { PensionArrangement } from '../../lib/types';
import { ContactDetails } from '../ContactDetails';
import { DefinitionList } from '../DefinitionList';

export const PensionArrangementCallout = (item: PensionArrangement) => {
  const { t, locale } = useTranslation();

  const listData = [...ContactDetails(item.pensionAdministrator)];

  return (
    <InformationCallout
      key={item.externalAssetId}
      className="px-4 py-6 pb-2 mt-6 md:py-7 md:px-5 md:mt-9"
    >
      <Heading
        level="h4"
        color="text-blue-700"
        className="flex gap-4 mb-4 md:mb-5"
      >
        <Icon
          type={IconType.WARNING_SQUARE}
          className="w-[30px] h-[30px] fill-red-700"
        />
        {item.schemeName}
      </Heading>

      <Paragraph className="md:mb-5">
        {t('pages.pensions-that-need-action.card.description', {
          pensionAdministrator: item.pensionAdministrator.name,
        })}
      </Paragraph>

      <div className="mb-3 md:mb-5">
        <Paragraph className="mb-0 font-bold">
          {t('pages.pensions-that-need-action.card.reference')}
        </Paragraph>

        <Paragraph className="mb-0" testId="pension-contact-reference">
          {item.contactReference ?? t('common.no-data')}
        </Paragraph>

        {(item.pensionType === PensionType.DC ||
          item.pensionType === PensionType.DB) && (
          <form method="POST">
            <input
              type="hidden"
              name="pensionId"
              value={item.externalAssetId}
            />
            <input type="hidden" name="locale" value={locale} />
            <input type="hidden" name="pensionType" value={item.pensionType} />
            <Button
              className="inline"
              formAction="/api/set-pension-id"
              data-testid="details-link"
              variant="link"
            >
              {t('pages.pensions-that-need-action.card.details-link')}
            </Button>
          </form>
        )}
      </div>

      <div className="py-2 mb:py-4 border-slate-400 border-t-1">
        <ExpandableSection
          title={t('pages.pensions-that-need-action.card.contact-heading-open')}
        >
          <DefinitionList
            className="!w-full !p-0 mb-0 bg-white rounded-none [&>dl>dt:last-of-type]:border-b-0 [&>dl>dd:last-of-type]:border-b-0"
            dlClassName="md:grid-cols-[36%_64%]"
            items={listData}
          />
        </ExpandableSection>
      </div>
    </InformationCallout>
  );
};
