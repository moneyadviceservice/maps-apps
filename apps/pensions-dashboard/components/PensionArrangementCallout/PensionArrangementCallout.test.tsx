import { render, screen } from '@testing-library/react';

import {
  PensionArrangement,
  PhoneNumber,
  PostalAddress,
} from '../../lib/types';
import { PensionArrangementCallout } from './PensionArrangementCallout';

import '@testing-library/jest-dom/extend-expect';

jest.mock('@maps-react/hooks/useTranslation', () => ({
  __esModule: true,
  default: () => ({
    t: (key: string) => key,
  }),
}));

const mockdata = {
  externalAssetId: '1',
  matchType: 'DEFN',
  schemeName: 'State Pension',
  pensionType: 'SP',
  contributionsFromMultipleEmployers: false,
  pensionAdministrator: {
    name: 'DWP',
    contactMethods: [
      {
        preferred: true,
        contactMethodDetails: {
          number: '+123 1111111111',
          usage: ['A', 'M'],
        },
      },
    ],
  },
  externalPensionPolicyId: 'policy123',
  contactReference: 'REF123',
} as PensionArrangement;

describe('PensionArrangementCallout', () => {
  it('renders correctly', () => {
    const { container } = render(<PensionArrangementCallout {...mockdata} />);
    expect(container).toMatchSnapshot();
  });

  it('renders the component with scheme name', () => {
    render(<PensionArrangementCallout {...mockdata} />);
    expect(screen.getByText('State Pension')).toBeInTheDocument();
  });

  it('renders the contact reference', () => {
    render(<PensionArrangementCallout {...mockdata} />);
    expect(screen.getByText('REF123')).toBeInTheDocument();
  });

  it('does not render the contact reference if undefined', () => {
    render(
      <PensionArrangementCallout
        {...{ ...mockdata, contactReference: undefined }}
      />,
    );
    expect(
      screen.queryByTestId('pension-contact-reference'),
    ).not.toBeInTheDocument();
  });

  it('renders the pension administrator name in the description', () => {
    render(<PensionArrangementCallout {...mockdata} />);
    expect(
      screen.getByText('pages.pensions-that-need-action.card.description'),
    ).toBeInTheDocument();
  });

  it('renders the preferred contact method', () => {
    render(<PensionArrangementCallout {...mockdata} />);
    expect(screen.getByText('common.contact.preferred')).toBeInTheDocument();
  });

  it('renders the telephone contact method', () => {
    const mockdataWithPhone = {
      ...mockdata,
      pensionAdministrator: {
        ...mockdata.pensionAdministrator,
        contactMethods: [
          ...mockdata.pensionAdministrator.contactMethods,
          {
            preferred: false,
            contactMethodDetails: {
              number: '+123 2222222222',
              usage: ['M'],
            } as PhoneNumber,
          },
        ],
      },
    };
    render(<PensionArrangementCallout {...mockdataWithPhone} />);
    expect(
      screen.getByText('Main telephone: +123 2222222222'),
    ).toBeInTheDocument();
  });

  it('renders the email contact method', () => {
    const mockdataWithEmail = {
      ...mockdata,
      pensionAdministrator: {
        ...mockdata.pensionAdministrator,
        contactMethods: [
          ...mockdata.pensionAdministrator.contactMethods,
          {
            preferred: false,
            contactMethodDetails: {
              email: 'admin@example.com',
            },
          },
        ],
      },
    };
    render(<PensionArrangementCallout {...mockdataWithEmail} />);
    expect(screen.getByText('admin@example.com')).toBeInTheDocument();
  });

  it('renders the web contact method', () => {
    const mockdataWithWebContact = {
      ...mockdata,
      pensionAdministrator: {
        ...mockdata.pensionAdministrator,
        contactMethods: [
          ...mockdata.pensionAdministrator.contactMethods,
          {
            preferred: false,
            contactMethodDetails: {
              url: 'https://example.com/contact',
            },
          },
        ],
      },
    };
    render(<PensionArrangementCallout {...mockdataWithWebContact} />);
    expect(screen.getByText('Contact DWP online')).toBeInTheDocument();
  });

  it('renders the postal address contact method', () => {
    const mockdataWithPostalAddress = {
      ...mockdata,
      pensionAdministrator: {
        ...mockdata.pensionAdministrator,
        contactMethods: [
          ...mockdata.pensionAdministrator.contactMethods,
          {
            preferred: false,
            contactMethodDetails: {
              postalName: 'DWP Office',
              line1: 'line1',
              line2: 'line2',
              line3: 'line3',
              line4: 'line4',
              line5: 'line5',
              postcode: '12345',
              countryCode: 'UK',
            } as PostalAddress,
          },
        ],
      },
    };
    render(<PensionArrangementCallout {...mockdataWithPostalAddress} />);
    expect(screen.getByText('DWP Office')).toBeInTheDocument();
    expect(screen.getByText('line1')).toBeInTheDocument();
    expect(screen.getByText('line2')).toBeInTheDocument();
    expect(screen.getByText('line3')).toBeInTheDocument();
    expect(screen.getByText('line4')).toBeInTheDocument();
    expect(screen.getByText('line5')).toBeInTheDocument();
    expect(screen.getByText('12345')).toBeInTheDocument();
  });
});
