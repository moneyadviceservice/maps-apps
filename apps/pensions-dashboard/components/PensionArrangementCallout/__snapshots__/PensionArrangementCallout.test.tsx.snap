// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PensionArrangementCallout renders correctly 1`] = `
<div>
  <div
    class="t-information-callout border border-slate-400 rounded-bl-3xl px-4 py-6 pb-2 mt-6 md:py-7 md:px-5 md:mt-9"
    data-testid="information-callout"
  >
    <h4
      class="text-xl md:text-2xl text-blue-700 font-bold flex gap-4 mb-4 md:mb-5"
    >
      <svgr-mock
        classname="w-[30px] h-[30px] fill-red-700"
      />
      State Pension
    </h4>
    <p
      class="mb-4 md:mb-5"
      data-testid="paragraph"
    >
      pages.pensions-that-need-action.card.description
    </p>
    <div
      class="mb-3 md:mb-5"
    >
      <p
        class="mb-0 font-bold"
        data-testid="paragraph"
      >
        pages.pensions-that-need-action.card.reference
      </p>
      <p
        class="mb-0"
        data-testid="pension-contact-reference"
      >
        REF123
      </p>
    </div>
    <div
      class="py-2 mb:py-4 border-slate-400 border-t-1"
    >
      <details
        class="group tool-expand"
        data-testid="expandable-section"
      >
        <summary
          class="text-magenta-500 flex [&::-webkit-details-marker]:hidden cursor-pointer py-2 outline-none focus:shadow-expander-focus underline hover:no-underline group"
        >
          <svgr-mock
            classname="group-open:rotate-[-180deg] text-magenta-500 shrink-0 w-5 mr-2"
          />
          <div
            class="block"
            data-testid="summary-block-title"
          >
            pages.pensions-that-need-action.card.contact-heading-open
          </div>
        </summary>
        <div
          class="mb-4 text-gray-800 text-base leading-[28px] tracking-[0.18px] pt-2"
        >
          <div
            class="p-4 pb-9 lg:p-14 lg:pt-13 lg:rounded-bl-[40px] xl:w-10/12 2xl:w-9/12 !w-full !p-0 mb-0 bg-white rounded-none [&>dl>dt:last-of-type]:border-b-0 [&>dl>dd:last-of-type]:border-b-0"
          >
            <dl
              class="md:grid md:grid-cols-[36%_64%]"
              data-testid="definition-list"
            >
              <dt
                class="block pt-2 pb-2 max-sm:leading-[1.6] md:border-b-1 md:border-slate-400 md:py-4 lg:pl-2 font-bold"
                data-testid="dt-contact-telephone"
              >
                common.contact.telephone
              </dt>
              <dd
                class="block pb-2 leading-[1.6] border-b-1 border-slate-400 md:py-4"
                data-testid="dd-contact-telephone"
              >
                <p
                  class="font-bold"
                >
                  (common.contact.preferred)
                   
                  WhatsApp: +123 1111111111
                </p>
              </dd>
            </dl>
          </div>
        </div>
      </details>
    </div>
  </div>
</div>
`;
